Basic website refactor and tutorials for models and mapping

dccowan opened the pull request:
This PR has two components:
1. A basic refactoring of the SimPEG website
2. Adding tutorials for generating models and defining mappings

**Basic Refactoring**

**The content of the website has now been parsed into several sections : getting started, tutorials, documentation and API. This is not the final form of the website, but it offers great improvement in the mean time.**

![image](https://user-images.githubusercontent.com/12970009/68970188-8e7b6c00-079b-11ea-8791-f8b274dba392.png)

**Mesh, model and mapping tutorials**

**We link to the mesh building tutorials on the discretize website. We also have tutorials for building models and defining mappings on all mesh types. When simulation branch is finalized, it will be trivial to adapt the syntax in these tutorials.**

![image](https://user-images.githubusercontent.com/12970009/68970339-dbf7d900-079b-11ea-96bb-a7f7e00f8ec7.png)

domfournier commented on tutorials/models_mapping/README.txt:
```diff
@@ -0,0 +1,12 @@
+Models and Mapping
+******************
+
+In SimPEG, the model represents the set of parameters that define our geology.
```
`defines`

domfournier commented on tutorials/models_mapping/README.txt:
```diff
@@ -0,0 +1,12 @@
+
+In SimPEG, the model represents the set of parameters that define our geology.
+For many problems, the model is a vector containing one or more physical
+property values for each cell in the mesh. However, we will see that there are
```
Remove last sentence.

domfournier replied:
I am thinking of a more direct approach where you list the topics/objectives of the tutorial, rather than a "story" type with specific use cases. Also, if we can find a way to hyperlink that would epic.. I really like this style (of course it's scipy... ):
https://docs.scipy.org/doc/scipy/reference/tutorial/interpolate.html

dccowan replied:
I changed the last sentence. But I like taking the time to explain *what* something is. For some users, this may be the first time they have seen certain concepts. I don't want to write an essay but I do want to offer some context. For individual tutorials, I think there are clearer objectives.

domfournier commented on tutorials/models_mapping/plot_1_tensor_models.py:
```diff
@@ -0,0 +1,337 @@
+
+
+"""
+Tensor Meshes
```
Singular Mesh

dccowan replied:
I think unless it is specified, most people would hear tensor mesh and assume there is only one mesh.

domfournier commented on tutorials/models_mapping/plot_1_tensor_models.py:
```diff
@@ -0,0 +1,337 @@
+# In this example we generate a half-space model. Since air cells remain
+# constant during geophysical inversion, the number of model values we define
+# should be equal to the number of cells lying below the surface. Here, we
+# define the model (*mod* ) as well as the mapping (*mod_map* ) that goes from
```
Remove this sentence and use plain English for variable names, even if it's long.

dccowan replied:
I changed the variable names, but I like giving context to what is happening.

domfournier commented on tutorials/models_mapping/plot_1_tensor_models.py:
```diff
@@ -0,0 +1,337 @@
+# Find cells below topography and define mapping
+air_val = 0.
+ind_active = mesh.gridCC[:, 2] < 0.
+mod_map = maps.InjectActiveCells(mesh, ind_active, air_val)
```
`model_map =`

dccowan replied:
done

domfournier commented on tutorials/models_mapping/plot_1_tensor_models.py:
```diff
@@ -0,0 +1,337 @@
+mod_map = maps.InjectActiveCells(mesh, ind_active, air_val)
+
+# Define the model on subsurface cells
+mod = background_val*np.ones(ind_active.sum())
```
`model = `

dccowan replied:
done

domfournier commented on tutorials/models_mapping/plot_1_tensor_models.py:
```diff
@@ -0,0 +1,337 @@
+mod_map = maps.InjectActiveCells(mesh, ind_active, air_val)
+
+# Define the model
+mod = halfspace_val*np.ones(ind_active.sum())
```
`model = `

dccowan replied:
done

domfournier commented on tutorials/models_mapping/plot_1_tensor_models.py:
```diff
@@ -0,0 +1,337 @@
+mod = background_val*np.ones(ind_active.sum())
+ind_dyke = (mesh.gridCC[ind_active, 0] > 20.) & (mesh.gridCC[ind_active, 0] < 40.)
+mod[ind_dyke] = dyke_val
+ind_block = (
```
Can we use `ModelBuilder.PolygonInd` here too?

dccowan replied:
I actually did some of this type of stuff deliberately. I know we are spending time developing great utilities. But I wanted to demonstrate that you could find the indicies in this way and change model parameters in this fashion.

lheagy commented on tutorials/models_mapping/plot_3_tree_models.py:
```diff
@@ -0,0 +1,371 @@
+mesh.plotSlice(
+    mod_map*mod, normal='Y', ax=ax, ind=int(mesh.hy.size/2), grid=True
+)
+ax.set_title('Model slice at y = -2.5 m')
```
and here

dccowan replied:
done

lheagy commented on tutorials/models_mapping/plot_3_tree_models.py:
```diff
@@ -0,0 +1,371 @@
+# ----------------------------
+#
+# Here we show how model building utilities are used to make more complicated
+# structural models. The process of adding a new unit is twofold: 1) we must
```
```suggestion
# structural models. The process of adding a new unit is twofold: (1) we must
```

lheagy commented on tutorials/models_mapping/plot_3_tree_models.py:
```diff
@@ -0,0 +1,371 @@
+#
+# Here we show how model building utilities are used to make more complicated
+# structural models. The process of adding a new unit is twofold: 1) we must
+# find the indicies for mesh cells that lie within the new unit, 2) we
```
```suggestion
# find the indicies for mesh cells that lie within the new unit, (2) we
```

lheagy commented on tutorials/models_mapping/plot_3_tree_models.py:
```diff
@@ -0,0 +1,371 @@
+mesh.plotSlice(
+    mod_map*mod, normal='Y', ax=ax, ind=int(mesh.hy.size/2), grid=True
+)
+ax.set_title('Model slice at y = -2.5 m')
```
here on title again

dccowan replied:
done

lheagy commented on tutorials/models_mapping/plot_3_tree_models.py:
```diff
@@ -0,0 +1,371 @@
+
+
+#############################################
+# Using Wire Maps
```
this is a great explanation @dccowan!

dccowan replied:
sweet

dccowan commented on tutorials/models_mapping/plot_3_tree_models.py:
```diff
@@ -0,0 +1,371 @@
+mesh.plotSlice(
+    mod_map*mod, normal='Y', ax=ax, ind=int(mesh.hy.size/2), grid=True
+)
+ax.set_title('Model slice at y = -2.5 m')
```
sure

dccowan commented on tutorials/models_mapping/plot_1_tensor_models.py:
```diff
@@ -0,0 +1,337 @@
+fig = plt.figure(figsize=(5, 5))
+ax = fig.add_subplot(111)
+mesh.plotSlice(sig_map*mod, normal='Y', ax=ax, ind=int(mesh.nCy/2), grid=True)
+ax.set_title('Model slice at y = -2.5 m')
```
Took care of this

dccowan commented:
@fourndo and @lheagy . I have looked over your comments and made changes. Variable names have been changed to be less concise but more descriptive. Labeling of plots is less hard coded. But I feel strongly about the current text and commenting for these tutorials. And several people outside our little bubble (Ren for instance) agree.

dccowan commented:
@lheagy This PR passed travis before making changes based on comments. Now it is failing. All I did was change the names of some variables and changed text. I could use some help finding the issue because the docs test doesn't offer much. I have done a 'make clean' and deleted everything autogenerated from the sphinx_gallery. I have a few theories right now:
- The current python environment doesn't like the proposed string formatting for my plots
- By deleting and regenerating the examples gallery, it forced travis to check if the examples would pass. And I'm not sure they do.
